fpdf/html.py

```diff
@@ -270,8 +270,8 @@ def __init__(
         self,
         pdf,
         image_map=None,
-        li_tag_indent=5,
-        dd_tag_indent=10,
+        li_tag_indent=30,
```

Comment:
Those are still magic numbers (overly large now for mm), which will vary in their effect depending on the current document units.
`pdf.k` contains the scale factor between the current document units and typographical points (pica), which the PDF format uses internally. To get the conversion factor between an arbitrary unit and pica, you can use eg, `util.get_scale_factor("mm")`.

You will also need test cases specifically for verifying that any constants defined internally really give the same result even when the user has set different document units.

Reply:
> `pdf.k` contains the scale factor between the current document units and typographical points (pica), which the PDF format uses internally. To get the conversion factor between an arbitrary unit and pica, you can use eg, `util.get_scale_factor("mm")`.

`pdf.k` is already defined using `util.get_scale_factor`, though.

> You will also need test cases specifically for verifying that any constants defined internally really give the same result even when the user has set different document units

I have already introduced `test_html_measurement_units`. Do I need to have tests for something not covered by it?

Reply:
> pdf.k is already defined using util.get_scale_factor, though.

Yes of course, when the user sets the document units, then unit `util.get_scale_factor()` is used to populate `self.k` with the appropriate conversion factor. But if we define built-in constants in mm, then we need to use it directly, to make sure we're independent of the document units. It's a bit tricky to keep track of what unit any specific value is in, so we need to to be careful and validate as much as possible.

> I have already introduced test_html_measurement_units

I didn't see that, will check it out.

Reply:
3 tests are failing:

test_html_customize_ul
test_html_li_tag_indent
test_html_measurement_units

Reply:
I don't have enough time to deal with things quite right now, so I will be tackling this stuff in a few hours instead.

> Yes of course, when the user sets the document units, then unit `util.get_scale_factor()` is used to populate `self.k` with the appropriate conversion factor. But if we define built-in constants in mm, then we need to use it directly, to make sure we're independent of the document units. It's a bit tricky to keep track of what unit any specific value is in, so we need to to be careful and validate as much as possible

Where should we get the information about the units from, in that case? Should I introduce a separate `FPDF.unit` attribute?

> Now I see. They look the same, which is good. Actually, you could check against the same file with all units, to verify that automatically

Any advice on how to do that elegantly?  
Changing `FPDF.k` seems inelegant, but I don't see another way of handling this without introducing changes to `FDPF`. So, I take it, I should introduce `FPDF.unit`?

> 3 tests are failing: ...

My bad. I didn't do the tests this time.

Reply:
> Where should we get the information about the units from, in that case? Should I introduce a separate FPDF.unit attribute?

That would just be a redundant copy of `pdf.k`. You can convert between all units if you use points as an intermediate value.
* From current units to points: `x * pdf.k`
* From mm to points: `x * util.get_scale_factor("mm")`
* From mm to current units: `x * util.get_scale_factor("mm") / pdf.k`
* etc.
It seems to me that you're actually giving the default values in points (30 pts ~= 12 mm). Is this intentional? 

>> Now I see. They look the same, which is good. Actually, you could check against the same file with all units, to verify that automatically
> 
> Any advice on how to do that elegantly?

Just use the same file name for all units. They should result in identical output, after all. (If you're very unlucky though, then rounding errors will make them just a tiny bit different...)

Reply:
> It seems to me that you're actually giving the default values in points (30 pts ~= 12 mm). Is this intentional?

Yes. This way the conversion is just a simple division by `self.pdf.k`. I am still unsure why we want to replace that with division by `util.get_scale_factor()`, and what to pass to the function as a parameter, unless I am to introduce `FPDF.unit`. I do question the utility of making additional function calls when we already store the relevant information in `self.pdf.k`.

> Just use the same file name for all units

The question is `how to best go about it?` Without storing information about units in `FPDF`, the solution seems to be to modify `pdf.k`, and, as you noted, there might be rounding errors.  
Introducing the `FPDF. unit` property attribute with a setter that also modifies `FPDF.k` introduces some redundancy. I have also encountered some issues with this approach - got misaligned indents and failure to generate pages that use centimetres or inches.

Reply:
> > It seems to me that you're actually giving the default values in points (30 pts ~= 12 mm). Is this intentional?
> 
> Yes. This way the conversion is just a simple division by `self.pdf.k`

Readability (for humans) is much more important than computational efficiency in most cases, and especially for a conversion that happens at most once for every table (in some cases once per program run). And where converting a constant is really too expensive, it needs a comment explaining what unit it is and what that means in real world terms.
Actually, any dimensional constant needs a comment specifying its unit anyway (unless the name already makes it clear).

> I am still unsure why we want to replace that with division by util.get_scale_factor()

I explained that several times already:
* `pdf.k` is the conversion factor for the current document units. You don't know what the user will supply for those (could be anything), and you don't want to know.
* You call `util.get_scale_factor(<unit>)` directly if you need the conversion factor for a fixed and known unit (for constants).

> and what to pass to the function as a parameter

Have a look at the definition of `util.get_scale_factor()`. That might make things more clear.

> > Just use the same file name for all units
> 
> The question is `how to best go about it?`

Just don't vary the file name for each iteration. There's no problem at all to have several tests compare against the same file.

> as you noted, there might be rounding errors.

Why do you let a hypothetical problem stop you from trying? This is a theoretical possibility, but since we only store a very limited number of decimal digits in the PDF file, differences through rounding errors are actually very rare (which is why I wrote "if you're very unlucky").

Reply:
> I explained that several times already:  
pdf.k is the conversion factor for the current document units. You don't know what the user will supply for those (could be anything), and you don't want to know.
You call `util.get_scale_factor(<unit>)` directly if you need the conversion factor for a fixed and known unit (for constants).

Do I understand it correctly, then, that I should introduce the attribute `FPDF.unit`, potentially as a property, and then convert default values of relevant variables with `* get_scale_factor('mm') / get_scale_factor(self.pdf.unit)`? This approach seems to currently do the same thing as `* get_scale_factor('mm') / self.pdf.k` while making this conversion potentially not cohere with the scale factor of the document itself. In addition, it seems to contradict what you said previously regarding the addition of the `FPDF.unit` attribute:

> That would just be a redundant copy of `pdf.k`

___

> Have a look at the definition of `util.get_scale_factor()`. That might make things more clear.

I have. We need to pass the chosen unit as a parameter, and currently there is no `FPDF.attribute`, meaning that to get the relevant unit, we would have to compare `FPDF.k` to possible return values of `utils.get_scale_factor()`.

For the time being, I will assume that I misunderstood something, return the default values of relevant variables to what they were before, and convert them to document units with `* get_scale_factor('mm') / self.pdf.k`. If need be, I will implement conversion using `get_scale factor()` instead of `self.pdf.k`.

> Why do you let a hypothetical problem stop you from trying?

Math brain, being used to preferring perfect precision.

Reply:
I repeat: You don't need and want to care about what the document units actually are (or how they might be named). There's a reason those are not actually stored anywhere. All you need is the conversion factor stored in `FPDF.k`. This was set once using `get_scale_factor(<user_supplied_unit>)`, and there's no point in repeating that.

You can use the same function to produce conversion factors for units of your own chosing.
Most size values in fpdf2 are stored and passed around in document units. So if you also store your constants in document units, that will save you many repeated conversions when they are used with other modules.

If you look at `FPDF.__init__()`, you'll find initializations like follows, converting the values from points, but with comments noting the mm/cm sizes:
```python
    self.line_width = 0.567 / self.k  # line width (0.2 mm)
    ...
    # Page margins (1 cm)
    margin = (7200 / 254) / self.k
```
You can follow this pattern for simplicity, or you can do somethin like this:
```python
    mm_to_doc_u = get_scale_factor("mm") / self.pdf.k 
    XXX = 123 * mm_to_doc_u  # this constant value
    YYY = 456 * mm_to_doc_u  # another constant value
```

> > Why do you let a hypothetical problem stop you from trying?
> 
> Math brain, being used to preferring perfect precision.

Alas, the real world is utterly imprecise... 😉 
The default resolution of the PDF format (eg. for images) is 72 dpi. We position elements more accurately than that, but any difference below 0.01 mm or so has no practical relevance, because it is imperceptible to the human eye.